cases/src/org/commcare/cases/instance/CaseDataInstance.java

```diff
@@ -0,0 +1,83 @@
+package org.commcare.cases.instance;
+
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.instance.utils.FormLoadingUtils;
+import org.javarosa.core.services.Logger;
+import org.javarosa.xml.util.InvalidStructureException;
+
+import java.io.IOException;
+
+/**
+ * An external data instance that respects CaseDB template specifications.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class CaseDataInstance extends ExternalDataInstance {
+    private static TreeElement caseDbSpecTemplate = null;
+    private static final String CASEDB_WILD_CARD = "CASEDB_WILD_CARD";
+    private static final Object lock = new Object();
+
+    public CaseDataInstance() {
+        // For serialization
+    }
+
+    /**
+     * Copy constructor
+     */
+    public CaseDataInstance(ExternalDataInstance instance) {
+        super(instance);
+    }
+
+    /**
+     * Does the reference follow the statically defined CaseDB spec?
+     */
+    public boolean hasTemplatePath(TreeReference ref) {
+        synchronized (lock) {
+            loadTemplateSpecLazily();
+
+            return followsTemplateSpec(ref, caseDbSpecTemplate, 0);
+        }
+    }
+
+    private static void loadTemplateSpecLazily() {
+        if (caseDbSpecTemplate == null) {
+            try {
+                caseDbSpecTemplate =
+                        FormLoadingUtils.xmlToTreeElement("/casedb_instance_structure.xml");
+            } catch (InvalidStructureException e) {
+                throw new RuntimeException("Failed to load casedb template spec");
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to load casedb template spec");
```

Comment:
Definitely agree. I might make this error very, very explanatory since I'm not sure what people would expect to happen here.